### PR TITLE
BUG in MultiIndex.drop for not-lexsorted multi-indexes, #12078

### LINKED ISSUE #12078: Obscur AttributeError when dropping on a multi-index dataframe
```
In [2]:  df = pd.DataFrame(columns=['a','b','c','d'], data=[[1,'b1','c1',3], [1,'b2','c2',4]])

In [3]: df = df.pivot_table(index='a', columns=['b','c'], values='d').reset_index()

In [4]: df
Out[4]: 
b  a b1 b2
c    c1 c2
0  1  3  4

In [5]: df.drop('a', axis=1)
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-5-b59fbf92d28f> in <module>()
----> 1 df.drop('a', axis=1)

/home/nicolas/Git/pandas/pandas/core/generic.pyc in drop(self, labels, axis, level, inplace, errors)
   1617                 new_axis = axis.drop(labels, level=level, errors=errors)
   1618             else:
-> 1619                 new_axis = axis.drop(labels, errors=errors)
   1620             dropped = self.reindex(**{axis_name: new_axis})
   1621             try:

/home/nicolas/Git/pandas/pandas/core/index.pyc in drop(self, labels, level, errors)
   5729                     inds.append(loc)
   5730                 else:
-> 5731                     inds.extend(lrange(loc.start, loc.stop))
   5732             except KeyError:
   5733                 if errors != 'ignore':

AttributeError: 'numpy.ndarray' object has no attribute 'start'
```

This is related to issue #11640. I have been working on a solution that I submitted in the pull request #11717, but the said solution was controversial, so I'm creating this issue to separate the problems.

I'll make a PR soon enough.

### PR DESCRIPTION
**Pull request by nbonnotte**
Closes #12078

### CODE REVIEW

File: pandas/tests/indexes/test_base.py [hunk 1]
```diff
@@ -1048,6 +1049,25 @@ def test_drop(self):
         expected = Index([1, 2])
         self.assert_index_equal(dropped, expected)
 
+    def test_drop_multiindex_not_lexsorted(self):
+        # GH 12078
```
Comment (jreback) [inline pandas/tests/indexes/test_base.py]: move to test_multi.py

File: pandas/tests/indexes/test_base.py [hunk 2]
```diff
@@ -1048,6 +1049,25 @@ def test_drop(self):
...
+        # define the lexsorted version of the multi-index
+        tuples = [('a', ''), ('b1', 'c1'), ('b2', 'c2')]
+        lexsorted_mi = MultiIndex.from_tuples(tuples, names=['b', 'c'])
+        self.assertTrue(lexsorted_mi.is_lexsorted())
+        # and the not-lexsorted version
```
Comment (jreback) [inline pandas/tests/indexes/test_base.py]: put blank line before a comment

File: pandas/indexes/multi.py [hunk 3]
```diff
@@ -1083,10 +1083,23 @@ def drop(self, labels, level=None, errors='raise'):
...
+                    if self.lexsort_depth == 0:
+                        warnings.warn('dropping on a non-lexsorted multi-index'
+                                      'without a level parameter may impact '
+                                      'performance.',
+                                      PerformanceWarning,
+                                      stacklevel=2)
+                    loc = loc.nonzero()[0]
+                    inds.extend(loc)
+                else:
+                    raise AssertionError
```
Comment (jreback) [inline pandas/indexes/multi.py]: put a nice message here (as we currently don't hit this), ``'unsupported indexer of type {0}'.format(type(loc))``

**Comment by nbonnotte** [comment]
I've taken into account the comments.

All green now.

**Comment by jreback** [comment]
merged via https://github.com/pydata/pandas/commit/f673af1bb51bb465144ffd2f843672e23433384a

thanks!

**Pull request closed without merging**